pkg/provider/azure_vmss.go

```diff
@@ -352,18 +370,29 @@ func (ss *ScaleSet) getVmssVMByInstanceID(resourceGroup, scaleSetName, instanceI
 // It must return ("", cloudprovider.InstanceNotFound) if the instance does
 // not exist or is no longer running.
 func (ss *ScaleSet) GetInstanceIDByNodeName(name string) (string, error) {
-	managedByAS, err := ss.isNodeManagedByAvailabilitySet(name, azcache.CacheReadTypeUnsafe)
+	vmManagementType, err := ss.getVMManagementTypeByNodeName(name, azcache.CacheReadTypeUnsafe)
 	if err != nil {
-		klog.Errorf("Failed to check isNodeManagedByAvailabilitySet: %v", err)
+		klog.Errorf("Failed to check VM management type: %v", err)
 		return "", err
 	}
-	if managedByAS {
+
+	if vmManagementType == ManagedByAvSet {
 		// vm is managed by availability set.
 		return ss.availabilitySet.GetInstanceIDByNodeName(name)
 	}
+	if vmManagementType == ManagedByVmssFlex {
+		// vm is managed by vmss flex.
+		return ss.flexScaleSet.GetInstanceIDByNodeName(name)
+	}
 
 	vm, err := ss.getVmssVM(name, azcache.CacheReadTypeUnsafe)
 	if err != nil {
+		// special case: during scaling in, if the vm is deleted and nonVmssUniformNodesCache is refreshed,
+		// then getVMManagementTypeByNodeName will return ManagedByVmssUniform no matter what the actual managementType is.
+		// In this case, if it is actually a non vmss uniform node, return InstanceNotFound
+		if errors.Is(err, ErrorNotVmssInstance) {
+			return "", cloudprovider.InstanceNotFound
```

Comment:
since the type has been checked above, why a node falls into this condition?

Reply:
This is based on a discovery of my testing. This applies to a special case:

Step 1: a scaling in operation deletes a node
Step 2: the nonVmssUniformNodesCache is refreshed by timeout

In the above case, the nonVmssUniformNodesCache does not contains the node name of this node  anymore so it assumes this is a uniform node. This is the case where it falls to this condition

Reply:
then is it possible for "scaling up" in step1 also falling into this condition? If so, `cloudprovider.InstanceNotFound` should not be returned here.

Are there any possible to trigger cache refreshing instead of default to "cloudprovider.InstanceNotFound" here?

Reply:
No. During scaling up, it first check if the node name exists in the cached cluster node names at the time of last cache refresh. It it does not exist, then it means this is a new node and we will trigger force cache refresh, so we don't need to worry about returning "cloudprovider.InstanceNotFound" in this case
<img width="1103" alt="image" src="https://user-images.githubusercontent.com/24629574/192864607-b8b4ead6-95fe-44d3-9213-60d0af28f572.png">